packages/edit-navigation/src/store/actions.js

```diff
@@ -163,57 +92,248 @@ export const saveNavigationPost = serializeProcessing( function* ( post ) {
 					saveError.message
 			  )
 			: __( 'Unable to save: An error ocurred.' );
-		yield dispatch( noticesStore, 'createErrorNotice', errorMessage, {
+		registry.dispatch( noticesStore ).createErrorNotice( errorMessage, {
 			type: 'snackbar',
 		} );
+	} finally {
+		registry.dispatch( coreDataStore ).__unstableReleaseStoreLock( lock );
 	}
-} );
+};
 
-function mapMenuItemsByClientId( menuItems, clientIdsByMenuId ) {
-	const result = {};
-	if ( ! menuItems || ! clientIdsByMenuId ) {
-		return result;
-	}
-	for ( const menuItem of menuItems ) {
-		const clientId = clientIdsByMenuId[ menuItem.id ];
-		if ( clientId ) {
-			result[ clientId ] = menuItem;
-		}
-	}
-	return result;
-}
+/**
+ * Executes appropriate insert, update, and delete operations to turn the current
+ * menu (with id=menuId) into one represented by the passed navigation block.
+ *
+ * @param {Object} navigationBlock The navigation block representing the desired state of the menu.
+ * @param {number} menuId          Menu Id to process.
+ * @return {Function} An action creator
+ */
+const batchSaveMenuItems = ( navigationBlock, menuId ) => async ( {
+	dispatch,
+	registry,
+} ) => {
+	// Make sure all the existing menu items are available before proceeding
+	const oldMenuItems = await registry
+		.resolveSelect( coreDataStore )
+		.getMenuItems( { menus: menuId, per_page: -1 } );
+
+	// Insert placeholders for new menu items to have an ID to work with.
+	// We need that in case these new items have any children. If so,
+	// we need to provide a parent id that we don't have yet.
+	const navBlockWithRecordIds = await dispatch(
+		batchInsertPlaceholderMenuItems( navigationBlock )
+	);
+
+	// Update menu items. This is separate from deleting, because there
+	// are no consistency guarantees and we don't want to delete something
+	// that was a parent node before another node takes it place.
+	const navBlockAfterUpdates = await dispatch(
+		batchUpdateMenuItems( navBlockWithRecordIds, menuId )
+	);
+
+	// Delete menu items
+	const deletedIds = difference(
+		oldMenuItems.map( ( { id } ) => id ),
+		blocksTreeToList( navBlockAfterUpdates ).map( getRecordIdFromBlock )
+	);
+	await dispatch( batchDeleteMenuItems( deletedIds ) );
+
+	return navBlockAfterUpdates;
+};
+
+/**
+ * Creates a menu item for every block that doesn't have an associated menuItem.
+ * Sends a batch request with one POST /wp/v2/menu-items for every created menu item.
+ *
+ * @param {Object} navigationBlock A navigation block to find created menu items in.
+ * @return {Function} An action creator
+ */
+const batchInsertPlaceholderMenuItems = ( navigationBlock ) => async ( {
+	registry,
+} ) => {
+	const blocksWithoutRecordId = blocksTreeToList( navigationBlock )
+		.filter( isSupportedBlock )
+		.filter( ( block ) => ! getRecordIdFromBlock( block ) );
```

Comment:
I think this could be collapsed to one filter call and it would still look readable. Something like this (I haven't tested this code):
```suggestion
	const blocksWithoutRecordId = blocksTreeToList( navigationBlock )
		.filter( ( block ) => isSupportedBlock( block ) && ! getRecordIdFromBlock( block ) );
```